modin/experimental/core/execution/native/implementations/omnisci_on_native/exchange/dataframe_protocol/column.py

```diff
@@ -0,0 +1,518 @@
+            return None
+
+        offset_buff = buffs[1]
+        # According to Arrow's data layout, the offset buffer type is "int32"
+        dtype = self._dtype_from_primitive_pandas(np.dtype("int32"))
+        return (
+            OmnisciProtocolBuffer(
+                offset_buff,
+                self._get_buffer_size(bit_width=dtype[1], is_offset_buffer=True),
+            ),
+            dtype,
+        )
+
+    def _propagate_dtype(self, dtype: Tuple[DTypeKind, int, str, str]):
+        """
+        Propagate `dtype` to the underlying PyArrow table.
+
+        Modifies the column object inplace by replacing underlying PyArrow table with
+        the casted one.
+
+        Parameters
+        ----------
+        dtype : tuple
+            Data type conforming protocol dtypes format to cast underlying PyArrow table.
+        """
+        if not self._col._allow_copy:
+            raise RuntimeError("Copy required with 'allow_copy=False' flag")
+
+        arrow_types_map = {
+            DTypeKind.BOOL: {8: pa.bool_()},
+            DTypeKind.INT: {
+                8: pa.int8(),
+                16: pa.int16(),
+                32: pa.int32(),
+                64: pa.int64(),
+            },
+            DTypeKind.UINT: {
+                8: pa.uint8(),
+                16: pa.uint16(),
+                32: pa.uint32(),
+                64: pa.uint64(),
+            },
+            DTypeKind.FLOAT: {16: pa.float16(), 32: pa.float32(), 64: pa.float64()},
+            DTypeKind.STRING: {8: pa.string()},
+        }
+        kind, bit_width, format_str, _ = dtype
+        arrow_type = None
+
+        if kind in arrow_types_map:
+            arrow_type = arrow_types_map[kind].get(bit_width, None)
+        elif kind == DTypeKind.DATETIME:
+            arrow_type = pa.timestamp("ns")
+        elif kind == DTypeKind.CATEGORICAL:
+            index_type = arrow_types_map[DTypeKind.INT].get(bit_width, None)
+            if index_type is not None:
+                arrow_type = pa.dictionary(
+                    index_type=index_type,
+                    # There is no way to deduce an actual value type, so casting to a string
+                    # as it's the most common one
+                    value_type=pa.string(),
+                )
+
+        if arrow_type is None:
+            raise NotImplementedError(f"Propagation for type {dtype} is not supported.")
+
+        at = self._pyarrow_table
+        schema_to_cast = at.schema
+        field = at.schema[0]
+
+        schema_to_cast = schema_to_cast.set(
+            0, pa.field(field.name, arrow_type, field.nullable)
+        )
+
+        # TODO: currently, each column chunk casts its buffers independently which results
+        # in an `N_CHUNKS - 1` amount of redundant casts. We can make the PyArrow table
+        # being shared across all the chunks, so the cast being triggered in a single chunk
+        # propagate to all of them.
+        self._cast_at(schema_to_cast)
```

Comment:
Each `ProtocolColumn` chunk has its own underlying PyArrow table, multiple PyArrow tables may internally use a single shared buffer (as a result of virtual chunking), however modifying a single table does not propagate changes to the rest of them, it just triggers a copy for this particular chunk to do the modifications.

When it's required to do PyArrow table casting for a single chunk, it's required for the rest of them as well. So the idea is to make a one single cast for the whole column at once, rather than casting each chunk independently. The benefits are: less memory allocations, less memory fragmentations.